src/spikeinterface/generation/hybrid_tools.py

```diff
@@ -0,0 +1,518 @@
+        if not favor_borders:
+            displacement *= rng.choice([-1.0, 1.0])
+            if unit_depth + displacement > top_margin:
+                displacement = -displacement
+            elif unit_depth - displacement < bottom_margin:
+                displacement = -displacement
+        else:
+            # check if depth is closer to top or bottom
+            if unit_depth > (top_margin - bottom_margin) / 2:
+                # if over top margin, move down
+                if unit_depth + displacement > top_margin:
+                    displacement = -displacement
+            else:
+                # if within bottom margin, move down
+                if unit_depth - displacement >= bottom_margin:
+                    displacement = -displacement
+        displacement_vector = np.zeros(2)
+        displacement_vector[depth_dimension] = displacement
+        templates_array_moved[i] = move_dense_templates(
+            templates.templates_array[i][None],
+            displacements=displacement_vector[None],
+            source_probe=templates.probe,
+        )[0]
+
+    return Templates(
+        templates_array=templates_array_moved,
+        sampling_frequency=templates.sampling_frequency,
+        nbefore=templates.nbefore,
+        sparsity_mask=templates.sparsity_mask,
+        channel_ids=templates.channel_ids,
+        unit_ids=templates.unit_ids,
+        probe=templates.probe,
+    )
+
+
+def generate_hybrid_recording(
+    recording,
+    motion_info=None,
+    num_units=10,
+    sorting=None,
+    templates=None,
+    templates_in_uV=True,
+    ms_before=1.0,
+    ms_after=3.0,
+    unit_locations=None,
+    upsample_factor=None,
+    upsample_vector=None,
+    amplitude_std=0.05,
+    generate_sorting_kwargs=dict(firing_rates=15, refractory_period_ms=4.0, seed=2205),
+    generate_unit_locations_kwargs=dict(margin_um=10.0, minimum_z=5.0, maximum_z=50.0, minimum_distance=20),
+    generate_templates_kwargs=dict(),
+    seed=None,
+):
+    """
+    Generate an hybrid recording with spike given sorting+templates.
+
+    Parameters
+    ----------
+    recording : BaseRecording
+        The recording to inject units in
+    motion_info : All the information about the motion
+        The motion datastructure of the recording
+    num_units : int, default: 10
+        Number of units,  not used when sorting is given.
+    sorting : Sorting or None
+        An external sorting object. If not provide, one is genrated.
+    templates : Templates or None, default: None
+        The templates of units.
+        If None they are generated.
+    templates_in_uV : bool, default: True
+        If True, the templates are in uV, otherwise they are in the same unit as the recording.
+        In case the recording has scaling, the templates are "unscaled" before injection.
+    ms_before : float, default: 1.5
+        Cut out in ms before spike peak.
+    ms_after : float, default: 3
+        Cut out in ms after spike peak.
```

Comment:
This should go in generate_template_kwargs no ?

Reply:
done